### PR TITLE
HSEARCH-899 Check for illegal usage of @IndexedEmbedded

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-899

### CODE REVIEW

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/mapping/definition/AnnotationMappingDiscoveryIT.java [hunk 1]
```diff
@@ -103,21 +105,7 @@ public void configure(HibernateOrmMappingConfigurationContext context) {
 
 	@Test
 	public void discoveryDisabled() {
-		backendMock.expectSchema( IndexedEntity.INDEX, b -> b
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/mapping/definition/AnnotationMappingDiscoveryIT.java]: You're completely changing the purpose of this test. Please don't do that.

This test is about checking that type annotation mapping discovery is effectively disabled. We used to check that by asserting that no mapping was contributed for the two `@IndexedEmbedded`. We can no longer do that because of the check you added.

One solution would be to add a programmatic mapping for the embedded types that adds some fields to these types, different from the ones configured through annotations. We would check that the programatically defined fields are added, but not the ones defined through annotations.

File: integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/AnnotationMappingDiscoveryIT.java [hunk 2]
```diff
@@ -93,22 +95,8 @@ public void discoveryEnabled() {
 
 	@Test
 	public void discoveryDisabled() {
-		backendMock.expectSchema( IndexedEntity.INDEX, b -> b
```
Comment (yrodiere) [inline integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/AnnotationMappingDiscoveryIT.java]: Same comment as `integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/mapping/definition/AnnotationMappingDiscoveryIT.java`. Please don't repurpose existing tests.

File: integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java [hunk 3]
```diff
@@ -657,18 +657,12 @@ public String getInvalid() {
...
 
-		backendMock.expectSchema( INDEX_NAME, b -> b
-				.field( "text", String.class )
-				// TODO HSEARCH-899 this isn't supposed to be legal => throw an exception for
-				.objectField( "invalid", c -> {
-				} )
-		);
-
-		JavaBeanMapping mapping = setupHelper.start()
+		SubTest.expectException( () -> setupHelper.start()
```
Comment (yrodiere) [inline integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java]: I think it would be clearer if you set these expectations right from the start, in the commit that adds the test. But to each his own, I guess.

File: integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java [hunk 4]
```diff
@@ -657,18 +657,12 @@ public String getInvalid() {
...
 				.withAnnotatedEntityTypes( IndexedEntity.class )
-				.setup();
-
-		backendMock.verifyExpectationsMet();
+				.setup() )
+				.assertThrown()
+				.isInstanceOf( SearchException.class )
+				.hasMessageContaining( "HSEARCH700060: The indexed embedded property 'invalid' of type 'java.lang.String' does not provide any valid field." );
```
Comment (yrodiere) [inline integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java]: Please don't mention the error code ("HSEARCH700060") in these assertions, they may change before the release (upon merging the PR, for example) and to be honest we don't really care about the exact code.

File: integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java [hunk 5]
```diff
@@ -708,18 +702,13 @@ public EmptyNested getInvalid() {
...
-		JavaBeanMapping mapping = setupHelper.start()
+		SubTest.expectException( () -> setupHelper.start()
 				.withAnnotatedEntityTypes( IndexedEntity.class )
-				.setup();
-
-		backendMock.verifyExpectationsMet();
+				.setup() )
+				.assertThrown()
+				.isInstanceOf( SearchException.class )
+				.hasMessageContaining( "HSEARCH700060: The indexed embedded property 'invalid' of type" )
```
Comment (yrodiere) [inline integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java]: Same comment here about the error code and context of the error.

File: mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/processing/building/impl/PojoIndexingProcessorTypeNodeBuilder.java [hunk 6]
```diff
@@ -135,13 +140,16 @@ public void closeOnFailure() {
...
+			if ( immutableBridgeHolders.isEmpty() && immutablePropertyNodes.isEmpty() ) {
+				if ( parentIndexObjectReferences.isEmpty() ) {
+					/*
+					 * If this node doesn't create any object in the document, and it doesn't have any bridge,
+					 * nor any property node, then it is useless and we don't need to build it.
+					 */
+					return Optional.empty();
+				}
+
+				throw log.invalidIndexedEmbedded( modelPath.getParent().getParent().getPropertyModel().getName(), modelPath.getTypeModel().getName() );
```
Comment (yrodiere) [inline mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/processing/building/impl/PojoIndexingProcessorTypeNodeBuilder.java]: This is not the right place for this error. I explicitly discussed the right way to implement this fix with you, and I mentioned why we needed to do it that way. I guess I wasn't clear enough.

Fields can be filtered out. Consider this:

```
@Entity 
@Indexed
public class Level1 {
    @IndexedEmbedded(includePaths = "level3b.text")
    private Level2 level2;
}

@Entity 
@Indexed
public class Level2 {
    @IndexedEmbedded
    private Level3A level3a;

    @IndexedEmbedded
    private Level3B level3b;
}

@Entity 
public class Level3A {
    @GenericField
    private String text;
}

@Entity 
public class Level3B {
    @GenericField
    private String text;
}
```

When building the indexing processor for `Level1`, it's entirely possible that a type node is created for `level3a`, but it won't define any field, because the `includePaths` clause filters it out.
Now this specific setup might not fail because there are optimizations in place which may trigger an early stop and prevent the type node from being created, but you get the idea: I'm sure there are other cases where an empty type node can legitimately be created.

The right way to address this problem, as we discussed previously, is to:

1. Replace `org.hibernate.search.engine.mapper.mapping.building.spi.TypeMetadataContributorProvider#forEach` with a method that takes a type model as a paremeter, and returns a set of `C`. Essentially `Set<C> get(MappableTypeModel typeModel)`.
2. In `org.hibernate.search.mapper.pojo.processing.building.impl.PojoIndexingProcessorValueNodeBuilderDelegate#indexedEmbedded`, test the returned set before doing anything: if it's empty, it means there isn't any mapping defined for the embedded type, and we should throw the exception.

File: mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/logging/impl/Log.java [hunk 7]
```diff
@@ -403,4 +403,7 @@ SearchException cannotPurgeNonIndexedContainedType(
 			value = "The bind() method of binder '%1$s' is not implemented correctly:"
 					+ " it did not call context.setMarker(...).")
 	SearchException missingMarkerForBinder(Object binder);
+
+	@Message(id = ID_OFFSET_2 + 60, value = "The indexed embedded property '%1$s' of type '%2$s' does not provide any valid field.")
+	SearchException invalidIndexedEmbedded(String propertyName, String propertyType);
```
Comment (yrodiere) [inline mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/logging/impl/Log.java]: Following the change in the implementation mentioned in my other comment, I'd rephrase the error like this:

```
	@Message(id = ID_OFFSET_2 + 60, value = "Type '%1$s' cannot be indexed-embedded, because no index mapping (@GenericField, @FullTextField, ...) is defined for that type.")
	SearchException invalidIndexedEmbedded(@FormatWith(PojoTypeModelFormatter.class) PojoTypeModel<?> typeModel); 
```

No need to mention the property name, it will be automatically included in the failure report (you can check that).

File: integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java (hunk 4 above)
Comment (yrodiere) [reply integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java]: Also, you'll probably want to check the context of the error (type, property name) in the failure report. See how we define expectations in `error_includePaths_nonMatched`.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/mapping/definition/AnnotationMappingDiscoveryIT.java (hunk 1 above)
Comment (fax4ever) [reply integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/mapping/definition/AnnotationMappingDiscoveryIT.java]: Good idea. Let's do that!

File: integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java (hunk 3 above)
Comment (fax4ever) [reply integrationtest/mapper/pojo/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/mapping/definition/IndexedEmbeddedBaseIT.java]: Squashed with the subsequent one.

File: mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/processing/building/impl/PojoIndexingProcessorTypeNodeBuilder.java (hunk 6 above)
Comment (fax4ever) [reply mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/processing/building/impl/PojoIndexingProcessorTypeNodeBuilder.java]: Sorry if I did remember that. Actually I don't have a good memory.
Not a problem to move the place where to address it.

Just for my very curiosity. How does it possible that:
> it won't define any field, because the includePaths clause filters it out.

if `Level3B` defines the generic field `text`?
Thanks ;)

Comment (fax4ever) [reply mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/processing/building/impl/PojoIndexingProcessorTypeNodeBuilder.java]: @yrodiere sorry. I'm trying to apply the strategy I think you pointed out here.

Probably I didn't grasp something. But the list is not empty, even if we have a String as a nested model type.

My attempt resides in the commit named:
`HSEARCH-899 Move check to PojoIndexingProcessorValueNodeBuilderDelegate`.

Thanks

**Comment by fax4ever** [comment]
@yrodiere thanks! 
Fixed the cases. Removing the odd '<no value extractors>' from the expected path context....
Is there anything else we can improve here?

File: mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/processing/building/impl/PojoIndexingProcessorTypeNodeBuilder.java (hunk 6 above)
Comment (yrodiere) [reply mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/processing/building/impl/PojoIndexingProcessorTypeNodeBuilder.java]: Looks like we solved this.

Regarding your question:

> Just for my very curiosity. How does it possible that:
>
> >    it won't define any field, because the includePaths clause filters it out.
>
> if Level3B defines the generic field text?

It's `level3a` which may end up empty, because it is not mentioned in `includePaths`. `level3b` and `level3b.text` will definitely be included, since they are explicitly mentioned in `includePaths`.

**Comment by yrodiere** [comment]
Looks good, merging. Thanks!
